core/trino-spi/src/main/java/io/trino/spi/type/SqlTimeWithTimeZone.java

```diff
@@ -35,6 +37,20 @@
 
     public static SqlTimeWithTimeZone newInstance(int precision, long picoseconds, int offsetMinutes)
     {
+        if (precision < 0 || precision > 12) {
+            throw new IllegalArgumentException("Invalid precision: " + precision);
+        }
+        if (rescale(rescale(picoseconds, 12, precision), precision, 12) != picoseconds) {
+            throw new IllegalArgumentException(format("picoseconds contains data beyond specified precision (%s): %s", precision, picoseconds));
+        }
+        if (picoseconds < 0 || picoseconds >= PICOSECONDS_PER_DAY) {
+            throw new IllegalArgumentException("picoseconds is out of range: " + picoseconds);
+        }
+        // TIME WITH TIME ZONE's valid offsets are [-14:00, 14:00]
```

Comment:
Did we enforce it already (just in a more ugly way during query processing). Or will some queries which used to work may start failing now?

Reply:
Good question. I checked literal form prior, but that's not enough. Filed https://github.com/trinodb/trino/issues/9288.